keras_hub/src/models/retinanet/box_matcher.py

```diff
@@ -0,0 +1,266 @@
+        iou_metric = keras_cv.bounding_box.compute_iou(anchors, boxes)
+        matched_columns, matched_match_values = box_matcher(iou_metric)
+        cls_mask = ops.less_equal(matched_match_values, 0)
+        ```
+
+    """
+
+    def __init__(
+        self,
+        thresholds,
+        match_values,
+        force_match_for_each_col=False,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        if sorted(thresholds) != thresholds:
+            raise ValueError(f"`threshold` must be sorted, got {thresholds}")
+        self.match_values = match_values
+        if len(match_values) != len(thresholds) + 1:
+            raise ValueError(
+                f"len(`match_values`) must be len(`thresholds`) + 1, got "
+                f"match_values {match_values}, thresholds {thresholds}"
+            )
+        thresholds.insert(0, -float("inf"))
+        thresholds.append(float("inf"))
+        self.thresholds = thresholds
+        self.force_match_for_each_col = force_match_for_each_col
+        self.built = True
+
+    def call(self, similarity_matrix):
+        """Matches each row to a column based on argmax
+
+        Args:
+            similarity_matrix: A float Tensor of shape `[num_rows, num_cols]` or
+                `[batch_size, num_rows, num_cols]` representing any similarity
+                metric.
+
+        Returns:
+            matched_columns: An integer tensor of shape `[num_rows]` or
+                `[batch_size, num_rows]` storing the index of the matched
+                column for each row.
+            matched_values: An integer tensor of shape [num_rows] or
+                `[batch_size, num_rows]` storing the match result
+                `(positive match, negative match, ignored match)`.
+        """
+        squeeze_result = False
+        if len(similarity_matrix.shape) == 2:
+            squeeze_result = True
+            similarity_matrix = ops.expand_dims(similarity_matrix, axis=0)
+        static_shape = list(similarity_matrix.shape)
+        num_rows = static_shape[1] or ops.shape(similarity_matrix)[1]
+        batch_size = static_shape[0] or ops.shape(similarity_matrix)[0]
+
+        def _match_when_cols_are_empty():
+            """Performs matching when the rows of similarity matrix are empty.
+            When the rows are empty, all detections are false positives. So we
+            return a tensor of -1's to indicate that the rows do not match to
+            any columns.
+
+            Returns:
+                matched_columns: An integer tensor of shape [batch_size,
+                    num_rows] storing the index of the matched column for each
+                    row.
+                matched_values: An integer tensor of shape [batch_size,
+                    num_rows] storing the match type indicator (e.g. positive or
+                    negative or ignored match).
+            """
+            with keras.name_scope("empty_boxes"):
```

Comment:
do we really need `name_scope` here? we have no usage of it elsewhere in KerasNLP, and it only does something on tensorflow. Probably we should remove this.

Reply:
Done